pkg/models/commandvalue.go

```diff
@@ -81,97 +91,115 @@ type CommandValue struct {
 // NewInt8Value creates a CommandValue of Type Int8 with the given value.
-func NewInt8Value(ro *models.ResourceOperation, origin int64, value int8) (cv *CommandValue, err error) {
+func NewInt8Value(ro *contract.ResourceOperation, origin int64, value int8) (cv *CommandValue, err error) {
 	cv = &CommandValue{RO: ro, Origin: origin, Type: Int8}
 	err = encodeValue(cv, value)
 	return
 }
 
 // NewInt16Value creates a CommandValue of Type Int16 with the given value.
-func NewInt16Value(ro *models.ResourceOperation, origin int64, value int16) (cv *CommandValue, err error) {
+func NewInt16Value(ro *contract.ResourceOperation, origin int64, value int16) (cv *CommandValue, err error) {
 	cv = &CommandValue{RO: ro, Origin: origin, Type: Int16}
 	err = encodeValue(cv, value)
 	return
 }
 
 // NewInt32Value creates a CommandValue of Type Int32 with the given value.
-func NewInt32Value(ro *models.ResourceOperation, origin int64, value int32) (cv *CommandValue, err error) {
+func NewInt32Value(ro *contract.ResourceOperation, origin int64, value int32) (cv *CommandValue, err error) {
 	cv = &CommandValue{RO: ro, Origin: origin, Type: Int32}
 	err = encodeValue(cv, value)
 	return
 }
 
 // NewInt64Value creates a CommandValue of Type Int64 with the given value.
-func NewInt64Value(ro *models.ResourceOperation, origin int64, value int64) (cv *CommandValue, err error) {
+func NewInt64Value(ro *contract.ResourceOperation, origin int64, value int64) (cv *CommandValue, err error) {
 	cv = &CommandValue{RO: ro, Origin: origin, Type: Int64}
 	err = encodeValue(cv, value)
 	return
 }
 
 // NewFloat32Value creates a CommandValue of Type Float32 with the given value.
-func NewFloat32Value(ro *models.ResourceOperation, origin int64, value float32) (cv *CommandValue, err error) {
+func NewFloat32Value(ro *contract.ResourceOperation, origin int64, value float32) (cv *CommandValue, err error) {
 	cv = &CommandValue{RO: ro, Origin: origin, Type: Float32}
 	err = encodeValue(cv, value)
 	return
 }
 
 // NewFloat64Value creates a CommandValue of Type Float64 with the given value.
-func NewFloat64Value(ro *models.ResourceOperation, origin int64, value float64) (cv *CommandValue, err error) {
+func NewFloat64Value(ro *contract.ResourceOperation, origin int64, value float64) (cv *CommandValue, err error) {
 	cv = &CommandValue{RO: ro, Origin: origin, Type: Float64}
 	err = encodeValue(cv, value)
 	return
 }
 
 //NewCommandValue create a CommandValue according to the Type supplied
-func NewCommandValue(ro *models.ResourceOperation, origin int64, value interface{}, t ValueType) (cv *CommandValue, err error) {
+func NewCommandValue(ro *contract.ResourceOperation, origin int64, value interface{}, t ValueType) (cv *CommandValue, err error) {
 	cv = &CommandValue{RO: ro, Origin: origin, Type: t}
-	if t != String {
-		err = encodeValue(cv, value)
-	} else {
-		cv.stringValue = value.(string)
+	switch t {
+		case Binary:
+			// assign cv.BinValue
+			err = encodeBinaryValue(cv, value)
```

Comment:
Do we need to encode here, or is the encoding of Events sufficient?

Reply:
Encoding of Events should be sufficient.
Will be addressed in a follow up PR.

Reply:
So after a close look, it appears the code is assuming that all parameters passed into to a write command are decoded as JSON, and then this function seems to encoding binary values into CBOR before being sent to the ProtocolDriver via handleWriteCommands().

I think what needs to happen is on a write command from a client, if one or more parameters are binary, then the whole request body should be CBOR encoded. So the write command logic should check the content header and decoded appropriately. The resulting binary data should be passed to the driver as a byte array.

Maybe it makes sense for handling write commands for this PR?

Reply:
@tonyespy That makes sense. I will add handling write commands to a new PR.